### PR TITLE
Unique signature indices

### PR DESCRIPTION
**Pull request by data-pup**
This PR adds the concept of unique signature indices to the `Module` struct in `wasmtime-environ`. Before, we were doing the dedup'ing in `crates/obj/src/context`. This moves the work up a little, and addresses what _was_ a `TODO` left in `crates/environ/src/module_environ.rs`.

### CODE REVIEW

File: crates/environ/src/module.rs [hunk 1]
```diff
@@ -12,6 +12,14 @@ use indexmap::IndexMap;
 use more_asserts::assert_ge;
 use std::hash::{Hash, Hasher};
 
+/// `UniqueSignatureIndex` names a signature after collapsing duplicate
+/// signatures to a single identifier, whereas SignatureIndex is directly
+/// what the original module specifies, and may specify duplicates of types
+/// that are structurally equal.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
+pub struct UniqueSignatureIndex(u32);
+entity_impl!(UniqueSignatureIndex);
```
Comment (data-pup) [inline crates/environ/src/module.rs]: Here's a new entity, for _unique_ signature indices.

File: crates/environ/src/module.rs [hunk 2]
```diff
@@ -133,8 +141,12 @@ impl TablePlan {
 // WARNING: when modifying, make sure that `hash_for_cache` is still valid!
 #[derive(Debug)]
 pub struct Module {
-    /// Unprocessed signatures exactly as provided by `declare_signature()`.
-    pub signatures: PrimaryMap<SignatureIndex, ir::Signature>,
+    /// This mapping lets us merge duplicate types (permitted by the wasm spec)
+    /// as they're declared.
+    pub signature_mapping: PrimaryMap<SignatureIndex, UniqueSignatureIndex>,
```
Comment (data-pup) [inline crates/environ/src/module.rs]: This map :point_up: from signature indices into unique signatures (used :point_down: there), is the important change to `Module` worth noting. Note that I haven't added any methods here, which _might_ be a good idea, depending on what reviewers think. I tried to be conservative for now.

File: crates/environ/src/module_environ.rs [hunk 3]
```diff
@@ -105,8 +105,15 @@ impl<'data> cranelift_wasm::ModuleEnvironment<'data> for ModuleEnvironment<'data
...
-        self.result.module.signatures.push(sig);
+        let unique_idx = self
+            .result
+            .module
+            .signatures
+            .iter()
+            .find(|(_, v)| **v == sig)
+            .map(|(k, _)| k)
+            .unwrap_or_else(|| self.result.module.signatures.push(sig));
+        self.result.module.signature_mapping.push(unique_idx);
```
Comment (data-pup) [inline crates/environ/src/module_environ.rs]: Here's where we de-duplicate signature declarations :slightly_smiling_face:

File: crates/obj/src/context.rs [hunk 4]
```diff
@@ -22,20 +19,12 @@ pub fn layout_vmcontext(
...
-        let target_index = match signature_registry.entry(sig) {
-            Entry::Occupied(o) => *o.get(),
-            Entry::Vacant(v) => {
-                assert_le!(signature_registry_len, std::u32::MAX as usize);
-                let id = TargetSharedSignatureIndex::new(signature_registry_len as u32);
-                signature_registry_len += 1;
-                *v.insert(id)
-            }
-        };
+    for (offset, target_index) in module.signatures.iter().map(|(index, _)| {
```
Comment (data-pup) [inline crates/obj/src/context.rs]: Previously, we were deduplicating our signatures here. Now that this is done with `ModuleEnvironment`, this code is simplified a little.

File: crates/jit/src/link.rs [hunk 5]
```diff
@@ -38,7 +38,11 @@ pub fn link_module(
                     signature,
                     vmctx,
                 } => {
-                    let import_signature = &module.signatures[module.functions[index]];
+                    let import_signature = {
+                        let sigidx = module.functions[index];
+                        let unique_index = module.signature_mapping[sigidx];
+                        &module.signatures[unique_index]
+                    };
```
Comment (data-pup) [inline crates/jit/src/link.rs]: One minor point: given a function index, getting a signature is a little more convoluted. I've used a small block scope for the sake of keeping these clearly tied to getting `import_signature`.

Comment (sunfishcode) [reply crates/jit/src/link.rs]: It looks like this sequence happens in a few different places. Could this code be factored into a helper method on `Module`?

**Comment by data-pup** [comment]
> Can you say more about the motivation for deduplicating signatures?
> [...] Are there runtime reasons for wanting to ensure a lack of duplicates?

I would like to try to start integrating lucet into wasmtime. In order to get there, I would like lucet to eventually use wasmtime's definition of a [wasm module](https://github.com/bytecodealliance/wasmtime/blob/39e57e3e9ac9c15bef45eb77a2544a7c0b76501a/crates/environ/src/module.rs#L135) , which looks fairly similar to what we have [in lucet](https://github.com/bytecodealliance/lucet/blob/master/lucetc/src/module.rs#L54).

Currently, lucet deduplicates both function and signature indices. It would be nice to hang onto these optimizations, so this is a small port of that logic to wasmtime.

As far as reasons to explicitly ensure a lack of duplicates, we would like to gracefully handle wasm modules that are not optimized. Using unique indices for functions and signatures also helped fix some other bugs in lucet related to [duplicate imports](https://github.com/bytecodealliance/lucet/pull/219) and [indirect calls](https://github.com/bytecodealliance/lucet/pull/168).

Admittedly I am newer to wasmtime's codebase, so feel free to let me know if this doesn't make sense to also have in `wasmtime`. My guess was that because signature indices were already being deduped when emitting an object file, it couldn't hurt to deduplicate earlier as well.

File: crates/jit/src/link.rs (hunk 5 above)
Comment (data-pup) [reply crates/jit/src/link.rs]: Yeah definitely! I'll take care of this.

**Comment by sunfishcode** [comment]
One of the reasons this is tricky is that Wasmtime does actually deduplicate signatures, but it does so at the level of the wasm store, rather than the wasm module. This allows it to support indirect calls across instance boundaries. So the second issue you linked to above, about indirect calls, is already fixed, just in a different way.

The other issue seems to be a case of some things deduplicating and not others. It's important to do things consistently, but it seems like we could go either way.

So, is it worthwhile to deduplicate signatures at the module level? It would make some tables smaller in some cases, but in my experience those tables aren't that big and don't have a lot of duplication in them. And it would make debugging a little more complex, because it's harder to see the correspondences between the in-memory `Module` and the actual module file. So my sense is that we're not ready to be doing these kinds of optimizations, at least without more evidence that they're important.

I am, of course, open to discussing it though.

**Comment by alexcrichton** [comment]
I ran across this today and this is pretty old at this point but I believe that the underlying feature here is actually implemented in Wasmtime nowadays already too. The main motivator was module linking, and I never got around to coming here and seeing this!

**Pull request closed without merging**